src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -494,67 +494,69 @@ private async Task ReleaseLeasesAsync(SqlConnection connection, CancellationToke
 
             try
             {
-                var transactionSw = Stopwatch.StartNew();
-                long releaseLeasesDurationMs = 0L, updateLastSyncVersionDurationMs = 0L;
-
-                using (SqlTransaction transaction = connection.BeginTransaction(System.Data.IsolationLevel.RepeatableRead))
+                for (int retryCount = 1; retryCount <= MaxRetryReleaseLeases; retryCount++)
                 {
-                    try
-                    {
-                        // Release the leases held on "_rows".
-                        using (SqlCommand releaseLeasesCommand = this.BuildReleaseLeasesCommand(connection, transaction))
-                        {
-                            this._logger.LogDebugWithThreadId($"BEGIN ReleaseLeases Query={releaseLeasesCommand.CommandText}");
-                            var commandSw = Stopwatch.StartNew();
-                            await releaseLeasesCommand.ExecuteNonQueryAsync(token);
-                            releaseLeasesDurationMs = commandSw.ElapsedMilliseconds;
-                            this._logger.LogDebugWithThreadId($"END ReleaseLeases Duration={releaseLeasesDurationMs}ms");
-                        }
+                    var transactionSw = Stopwatch.StartNew();
+                    long releaseLeasesDurationMs = 0L, updateLastSyncVersionDurationMs = 0L;
 
-                        // Update the global state table if we have processed all changes with ChangeVersion <= newLastSyncVersion,
-                        // and clean up the leases table to remove all rows with ChangeVersion <= newLastSyncVersion.
-                        using (SqlCommand updateTablesPostInvocationCommand = this.BuildUpdateTablesPostInvocation(connection, transaction, newLastSyncVersion))
+                    using (SqlTransaction transaction = connection.BeginTransaction(System.Data.IsolationLevel.RepeatableRead))
```

Comment:
One thing I've been thinking about that we need to be careful of is our use of transactions across the same connection. A couple things I've found looking through the docs and playing around with it myself : 

1. SqlConnection doesn't like it when multiple transactions are opened for the same connection. Although oddly this doesn't seem to matter with multiple threads, I was able to open multiple transactions as long as that was done on another thread...not sure if that's intended or not
2. If you try to execute a command on a connection with an open transaction it'll throw

Hopefully this shouldn't be a huge issue since we're opening a new connection for every worker that's processing changes, but we need to make sure that each worker isn't going to be doing anything in parallel that could cause issues.

Reply:
Thank for the insights. Here, we are having two loops running in parallel, changed-consumption and lease-renewal, and both use their own connections so no issue there. We will probably be adding a new loop for scale monitoring. I will ensure another connection is created for that one. As it seems, with every new change, we should ensure that the code is correctly committing (or rolling back) the transaction before beginning a new one.

Reply:
FYI - got a response on the issue I submitted

https://github.com/dotnet/SqlClient/issues/1758#issuecomment-1243072330

So we need to be careful that we're never opening a connection and then accessing it in another thread. I looked through everything and it seems like we're doing that currently everywhere so we should be safe but if you could also take a look through and make sure I'm not missing anything I'd appreciate it.